src/diversity/abundance.py

```diff
@@ -133,122 +83,62 @@ def normalized_subcommunity_abundance(self):
         pass
 
 
-class Abundance(IAbundance):
-    """Implements IAbundance for fast, but memory-heavy calculations.
+class AbundanceFromArray(Abundance):
+    """Implements Abundance for fast, but memory-heavy calculations.
 
     Caches counts and (normalized) relative meta- and subcommunity
     abundances at the same time.
     """
 
-    def __init__(self, counts):
-        """Initializes object.
-
+    def __init__(self, counts: Union[DataFrame, ndarray]) -> None:
+        """
         Parameters
         ----------
-        counts: numpy.ndarray
-            A 2-d numpy.ndarray with one column per subcommunity, one
-            row per species, containing the count of each species in the
-            corresponding subcommunities.
-
+        counts:
+            2-d array with one column per subcommunity, one row per species,
+            containing the count of each species in the corresponding subcommunities.
         """
         LOGGER.debug("Abundance(counts=%s", counts)
-        self.counts = counts
+        self.counts = DataFrame(counts)
 
-    @cache
-    def subcommunity_abundance(self):
-        total_abundance = self.counts.sum()
-        relative_abundances = empty(shape=self.counts.shape, dtype=dtype("f8"))
+    @cached_property
+    def subcommunity_abundance(self) -> ndarray:
+        total_abundance = self.counts.sum().sum()
+        relative_abundances = empty(shape=self.counts.shape, dtype=float64)
         relative_abundances[:] = self.counts / total_abundance
         return relative_abundances
 
-    @cache
-    def metacommunity_abundance(self):
-        return self.subcommunity_abundance().sum(axis=1, keepdims=True)
+    @cached_property
+    def metacommunity_abundance(self) -> ndarray:
+        return self.subcommunity_abundance.sum(axis=1, keepdims=True)
 
-    @cache
-    def subcommunity_normalizing_constants(self):
-        return self.subcommunity_abundance().sum(axis=0)
+    @cached_property
+    def subcommunity_normalizing_constants(self) -> ndarray:
+        return self.subcommunity_abundance.sum(axis=0)
 
-    @cache
-    def normalized_subcommunity_abundance(self):
-        return self.subcommunity_abundance() / self.subcommunity_normalizing_constants()
+    @cached_property
+    def normalized_subcommunity_abundance(self) -> ndarray:
+        return self.subcommunity_abundance / self.subcommunity_normalizing_constants
 
 
-class SharedAbundance(IAbundance):
```

Comment:
I have added a class `AbundanceFromSparseArray` which allows for the abundance arrays to be a scipy sparse arrays. This covers the case when the abundance is the identity matrix and other very sparse matrices (e.g. bag of words matrix). Now P and P_bar are sparse. Unfortunately, ZP and ZP_bar will still always be dense.

We could use the same trick of normalizing and un-normalizing while overwriting the same memory block, but this feel like unintuitive behavior (i.e., breaks the principle of least astonishment). For example, this could lead to more computations if the user swapped between normalized and un-normalized measures repeatedly in a loop. Worst of all, they wouldn't be aware that this is happening. Ultimately, it's a trade-off between memory or computation/code clarity.

I think that the code already covers many more big-data applications than the R and Julia package could, and with the addition of sparse array this covers an even greater set of use cases, so I am ok with taking the memory hit.

As a last note, `AbundanceFromSpareArray` requires that scipy's sparse array interface is used and not the sparse matrix interface. This is because the [scipy docs](https://docs.scipy.org/doc/scipy/reference/sparse.html) recommend all future work transition to the array interface. Maybe this is worth noting in our README/docs.